src/ImageSharp/Processing/Processors/Quantization/EuclideanPixelMap{TPixel}.cs

```diff
@@ -162,96 +128,415 @@ private int GetClosestColorSlow(Rgba32 rgba, ref TPixel paletteRef, out TPixel m
-        private const int AShift = 8 - IndexABits;
-        private const int Entries = IndexRCount * IndexGCount * IndexBCount * IndexACount;
-        private MemoryHandle tableHandle;
-        private readonly IMemoryOwner<short> table;
-        private readonly short* tablePointer;
-
-        public ColorDistanceCache(MemoryAllocator allocator)
+        private readonly CoarseCache coarseCache;
+        private readonly ExactCache exactCache;
+
+        public HybridColorDistanceCache(MemoryAllocator allocator)
+        {
+            this.exactCache = new ExactCache(allocator);
+            this.coarseCache = new CoarseCache(allocator);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public readonly void Add(Rgba32 color, short index)
         {
-            this.table = allocator.Allocate<short>(Entries);
-            this.table.GetSpan().Fill(-1);
-            this.tableHandle = this.table.Memory.Pin();
-            this.tablePointer = (short*)this.tableHandle.Pointer;
+            if (this.exactCache.TryAdd(color.PackedValue, index))
+            {
+                return;
+            }
+
+            this.coarseCache.Add(color, index);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public readonly bool TryGetValue(Rgba32 color, out short match)
+        {
+            if (this.exactCache.TryGetValue(color.PackedValue, out match))
+            {
+                return true; // Exact match found
+            }
+
+            if (this.coarseCache.TryGetValue(color, out match))
+            {
+                return true; // Coarse match found
+            }
+
+            match = -1;
+            return false;
+        }
+
+        public readonly void Clear()
+        {
+            this.exactCache.Clear();
+            this.coarseCache.Clear();
+        }
+
+        public void Dispose()
+        {
+            this.exactCache.Dispose();
+            this.coarseCache.Dispose();
+        }
+    }
+
+    /// <summary>
+    /// A fixed-capacity dictionary with exactly 512 entries mapping a <see cref="uint"/> key
+    /// to a <see cref="short"/> value.
+    /// </summary>
+    /// <remarks>
+    /// The dictionary is implemented using a fixed array of 512 buckets and an entries array
+    /// of the same size. The bucket for a key is computed as (key &amp; 0x1FF), and collisions are
+    /// resolved through a linked chain stored in the <see cref="Entry.Next"/> field.
+    /// The overall memory usage is approximately 4–5 KB. Both lookup and insertion operations are,
+    /// on average, O(1) since the bucket is determined via a simple bitmask and collision chains are
+    /// typically very short; in the worst-case, the number of iterations is bounded by 256.
+    /// This guarantees highly efficient and predictable performance for small, fixed-size color palettes.
+    /// </remarks>
+    internal sealed unsafe class ExactCache : IDisposable
```

Comment:
@rickbrew @saucecontrol Here's the new Exact and Coarse caches I built. They work really well!